lib/http2/connection.c

```diff
@@ -81,63 +79,58 @@ static void enqueue_goaway(h2o_http2_conn_t *conn, int errnum, h2o_iovec_t addit
 
 static void graceful_shutdown_close_stragglers(h2o_timer_t *entry)
 {
-    h2o_context_t *ctx = H2O_STRUCT_FROM_MEMBER(h2o_context_t, http2._graceful_shutdown_timeout, entry);
-    h2o_linklist_t *node, *next;
-
+    h2o_http2_conn_t *conn = H2O_STRUCT_FROM_MEMBER(h2o_http2_conn_t, _graceful_shutdown_timeout, entry);
     /* We've sent two GOAWAY frames, close the remaining connections */
-    for (node = ctx->http2._conns.next; node != &ctx->http2._conns; node = next) {
-        h2o_http2_conn_t *conn = H2O_STRUCT_FROM_MEMBER(h2o_http2_conn_t, _conns, node);
-        next = node->next;
-        close_connection(conn);
-    }
+    close_connection(conn);
 }
 
 static void graceful_shutdown_resend_goaway(h2o_timer_t *entry)
 {
-    h2o_context_t *ctx = H2O_STRUCT_FROM_MEMBER(h2o_context_t, http2._graceful_shutdown_timeout, entry);
-    h2o_linklist_t *node;
-    int do_close_stragglers = 0;
+    h2o_http2_conn_t *conn = H2O_STRUCT_FROM_MEMBER(h2o_http2_conn_t, _graceful_shutdown_timeout, entry);
+
+    if (conn->state < H2O_HTTP2_CONN_STATE_HALF_CLOSED) {
+        enqueue_goaway(conn, H2O_HTTP2_ERROR_NONE, (h2o_iovec_t){NULL});
 
-    for (node = ctx->http2._conns.next; node != &ctx->http2._conns; node = node->next) {
-        h2o_http2_conn_t *conn = H2O_STRUCT_FROM_MEMBER(h2o_http2_conn_t, _conns, node);
-        if (conn->state < H2O_HTTP2_CONN_STATE_HALF_CLOSED) {
-            enqueue_goaway(conn, H2O_HTTP2_ERROR_NONE, (h2o_iovec_t){NULL});
-            do_close_stragglers = 1;
+        /* After waiting a second, we still had active connections. If configured, wait one
+         * final timeout before closing the connections */
+        if (conn->super.ctx->globalconf->http2.graceful_shutdown_timeout > 0) {
+            conn->_graceful_shutdown_timeout.cb = graceful_shutdown_close_stragglers;
+            h2o_timer_link(conn->super.ctx->loop, conn->super.ctx->globalconf->http2.graceful_shutdown_timeout,
+                           &conn->_graceful_shutdown_timeout);
         }
     }
+}
 
-    /* After waiting a second, we still had active connections. If configured, wait one
-     * final timeout before closing the connections */
-    if (do_close_stragglers && ctx->globalconf->http2.graceful_shutdown_timeout > 0) {
-        ctx->http2._graceful_shutdown_timeout.cb = graceful_shutdown_close_stragglers;
-        h2o_timer_link(ctx->loop, ctx->globalconf->http2.graceful_shutdown_timeout, &ctx->http2._graceful_shutdown_timeout);
-    }
+static int close_idle_connection(h2o_conn_t *_conn)
+{
+    h2o_http2_conn_t *conn = (void *)_conn;
+    enqueue_goaway(conn, H2O_HTTP2_ERROR_NONE, h2o_iovec_init(H2O_STRLIT("idle timeout")));
+    close_connection(conn);
```

Comment:
We could act upon this immediately by adding a `h2o_quic_send(&conn->h3.super);` here, else the connection is marked as closed and a timeout will clean it later

Reply:
I'm not sure if want to do an immediate close in `close_idle_function`.

The intent of the function is to gracefully reduce amount of resource used, so that there would be more room for new connections. That's why we try to only close connections that are "idle."

To paraphrase, we do not want to close connections that look idle to us, but on which a request is actually inflight from the client. That's because doing so would disrupt the request inflight, and/or lead to the client reestablish a connection to resend the request. That'd be a lose-lose scenario to both ends.

I think the correct behavior here would be to start the graceful shutdown process by sending GOAWAY. The same argument applies to http3.

(and I recognize that we might need three linklists rather than two, because we'd need a separate linklist for connections in the graceful-shutdown state).